### PR TITLE
Add primops to access the current diagnostic of a label

### PR DESCRIPTION
**Pull request by yannham**
A previous PR (#1152) introduced a more fine-grained interface for contract labels, in order to customize error diagnostic. This PR add the Nickel primops to make this interface accessible to usercode, as well as the corresponding stdlib functions.

## Content

- Adds `label_with_message`, `label_with_notes`, `label_append_note`, `label_push_diag` primops, calling the corresponding methods on `Label`
- Adds corresponding functions in the stdlib, under the `contract.label` submodule
- Gets rid of `%tag%` and `contract.tag`, superseded by `%label_with_message%` and `contract.label.with_message`
- Renames accordingly `contract.blame_with` to `contract.blame_with_message`
- Makes `contract.apply` to automatically call `%label_push_diag%`. This means that `contract.apply` becomes the implicit split point between a parent contract and children subcontracts: it will stack any previously set custom diagnostic, which will save it from further modifications. Later calls to e.g. `contract.label.with_error_message` will modify a new, fresh diagnostic. All stacked diagnostics are printed during error reporting, in reverse order.

### CODE REVIEW

File: src/label.rs [hunk 1]
```diff
@@ -430,47 +430,62 @@ impl Label {
...
 
-    /// Return a reference to the current contract diagnostic, that is the last one of the stack,
-    /// if any.
+    /// Return a reference to the current contract diagnostic, which is the last element of the
+    /// stack, if any.
     pub fn current_diagnostic(&self) -> Option<&ContractDiagnostic> {
         self.diagnostics.last()
     }
+
+    /// Push a new fresh diagnostic on the stack if the current diagnostic isn't empty. This has
```
Comment (ebresafegaga) [inline src/label.rs:481]:
```suggestion
    /// Push a new, fresh diagnostic on the stack if the current diagnostic isn't empty. This has
```

**Review by ebresafegaga — APPROVED** [review]
Looks good 🤩

File: stdlib/contract.ncl [hunk 2]
```diff
@@ -5,103 +5,249 @@
...
+          anymore. All diagnostics will be shown during error reporting, with
+          the most recent being used as the main one.
+
+          `contract.apply` is the operation that pushes a new fresh diagnostic on
+          the stack, saving the previously current error diagnostic. Indeed,
+          `contract.apply` is mostly used to apply subcontracts inside a parent
+          contract. Stacking the current diagnostic potentially customized by
+          the parent contract saves the information inside, a provide a fresh
+          diagnostic for the child contract to use.
         "%
```
Comment (vkleen) [inline stdlib/contract.ncl:93]:
```suggestion
          the parent contract saves the information inside, and provides a fresh
          diagnostic for the child contract to use.
        "%
```

File: stdlib/contract.ncl [hunk 3]
```diff
@@ -5,103 +5,249 @@
...
-            value in
-        5 | Contract
-        ```
+          The label submodule, which gathers functions that manipulate the label
+          of a contract.
+
+          A label is a special opaque value automatically passed by the Nickel
+          interpreter to contracts when performing a contract check.
+
+          A label stores a stack of custom error diagnostics, than can be
```
Comment (vkleen) [inline stdlib/contract.ncl:76]:
```suggestion
          A label stores a stack of custom error diagnostics, that can be
```

File: stdlib/contract.ncl [hunk 4]
```diff
@@ -5,103 +5,249 @@
...
+          ({foo = 1} | Contract)
+          ```
+
+          # Diagnostic stack
+
+          Using `apply` will stack the current custom reporting data, and create a
+          fresh current working diagnostic. `apply` thus acts automatically as a
+          split point between a contract and its subcontracts, providing the
+          subcontracts with a fresh diagnostic to use, while remembering the
+          previous the previous diagnostics set by parent contracts.
```
Comment (vkleen) [inline stdlib/contract.ncl:222]:
```suggestion
          previous diagnostics set by parent contracts.
```

File: stdlib/contract.ncl [hunk 5]
```diff
@@ -5,103 +5,249 @@
...
+              |> contract.label.with_message "parent's message"
+              |> contract.label.append_note "parent's note"
+            in
+            contract.apply ChildContract label value
+          in
+
+          null | ParentContract
+          ```
+
+          This example will print two diagnostics: the main one using, the
```
Comment (vkleen) [inline stdlib/contract.ncl:246]:
```suggestion
          This example will print two diagnostics: the main one using the
```

**Review by vkleen — APPROVED** [review]
👍 No technical concerns, just some typochecking.

**Comment by yannham** [comment]
> typochecking

_Noice_ :sunglasses:

**Pull request merged**